python_benchmarks/core.py

```diff
@@ -0,0 +1,111 @@
+import sys
+import ctypes
+import torch
+from torch.profiler import profile, ProfilerActivity
+
+libnames = ("libcuda.so", "libcuda.dylib", "nvcuda.dll", "cuda.dll")
+for libname in libnames:
+    try:
+        cuda = ctypes.CDLL(libname)
+    except OSError:
+        continue
+    else:
+        break
+else:
+    raise OSError("could not load any of: " + " ".join(libnames))
+
+CU_DEVICE_ATTRIBUTE_L2_CACHE_SIZE = 38
+l2_cache_size = ctypes.c_int()
+device = torch.cuda.current_device()
+cuda.cuDeviceGetAttribute(
+    ctypes.byref(l2_cache_size), CU_DEVICE_ATTRIBUTE_L2_CACHE_SIZE, device
+)
+
+L2_CACHE_SIZE = l2_cache_size.value
+
+
+def clearL2Cache():
+    n_elements = L2_CACHE_SIZE // 4
+    x = torch.empty(n_elements, dtype=torch.float32, device="cuda", requires_grad=False)
+    y = torch.clone(x)
+
+
+class NVFBenchmark(object):
+    def __init__(self, benchmark_fixture, precision: float = 1e-6):
+        # Initialize a Torch Profiler object
+        self.prof = profile(activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU])
+
+        # Set the timer and precision
+        benchmark_fixture._timer = self.torchprofile_timer
+        benchmark_fixture._precisions[benchmark_fixture._timer] = precision
+
+        self.benchmark = benchmark_fixture
+        self.current_time = 0.0
+
+    def __call__(self, function_to_benchmark, *args, **kwargs):
+        return self.benchmark(function_to_benchmark, *args, **kwargs)
+
+    def __getattr__(self, attr):
+        if attr not in self.__dict__:
+            return getattr(self.benchmark, attr)
+        return super().__getattr__(attr)
+
+    def torchprofile_timer(self):
+        try:
+            self.prof.stop()
+            prof_output = self.prof.key_averages().table()
```

Comment:
What's the return type of `key_averages`? I'm surprised it doesn't have a more programmatic way to retrieve the results. Parsing stdout is fragile -- pytorch can change the output format anytime without notifying anyone.

Reply:
Modified this to directly compute the total time for CUDA events.